File Types Order & Type Contents Order

Dschee opened the pull request:
This fixes #2294 by introducing two new rules:
- `file_types_order`: Specifies how the types within a file should be ordered.
- `type_contents_order`: Specifies the order of subtypes, properties, methods & more within a type.

Dschee commented:
I just added configuration files to both rules, this means the only missing thing is fixing the locations in the tests. Since I have no idea why this happens, help is highly appreciated! @jpsim @marcelofabri @norio-nomura

marcelofabri commented:
> As you can see the offset to the init method points to somewhere right in the middle of the override keyword, which is weird.

I think it actually points to `init`, because we remove the violation marker from the string before passing to the rule.

Dschee commented:
@marcelofabri Okay, you were right about that example which I fixed now. But that was not what was irritating me, here's a few examples which now still are a problem and I can't understand why:

```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
↓class TestViewController: UIViewController {}
                                          ^ warning: File Types Order Violation: A 'main_type' should not be placed amongst the file type(s) 'supporting_type'. (file_types_order)

// Supporting Types
protocol TestViewControllerDelegate {
    func didPressTrackedButton()
}
```

```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
// Supporting Types
protocol TestViewControllerDelegate {
    func didPressTrackedButton()
}

// Extensions
↓extension TestViewController: UITableViewDataSource {
                                          ^ warning: File Types Order Violation: A 'extension' should not be placed amongst the file type(s) 'main_type'. (file_types_order)
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return UITableViewCell()
    }
}

class TestViewController: UIViewController {}

// Extensions
extension TestViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return UITableViewCell()
    }
}
```

```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
class TestViewController: UIViewController {
    // Subtypes
    ↓class TestClass {
        // 10 lines
    }
    ^ warning: Type Contents Order Violation: A 'subtype' should not be placed amongst the type content(s) 'type_alias,associated_type'. (type_contents_order)

    // Type Aliases
    typealias CompletionHandler = ((TestEnum) -> Void)
}
```

```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
class TestViewController: UIViewController {
    // IBActions
    @IBAction ↓func goNextButtonPressed() {
        goToNextVc()
             ^ warning: Type Contents Order Violation: A 'ib_action' should not be placed amongst the type content(s) 'view_life_cycle_method'. (type_contents_order)
        delegate?.didPressTrackedButton()
    }

    // View Life-Cycle Methods
    override func viewDidLoad() {
        super.viewDidLoad()

        view1.setNeedsLayout()
        view1.layoutIfNeeded()
        hasLayoutedView1 = true
    }
}
```

It only happens when there's the `/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */` prepended at the beginning.

Dschee commented:
Another problem I found was that when I uninstall SwiftLint entirely from my computer and run `make install` on my fork, then the new rules seem to work and find violations, but the `reason` statement isn't shown. As you can see in the examples above I have added specific reasons to explain what is placed amongst which things. This seems to work in tests, but when I compile the code into the CLI tool then in my project the violations simply state the `description` of the `RuleDescription`, not the specific `reason`. Is there something I need to do before this works?

Dschee commented:
I just fixed all issues and rebased this to the lastest master. Ready to merge from my point of view!

Dschee commented:
Okay, I just found the issue which made the CI fail when running this with Swift 4: It's due to `subscript` and `typealias` missing from SourceKit for versions below 4.1. I fixed this issue by not testing the cases which include the supporting type on Swift 4.

I hope the CI will be happy now and this can be merged. Anything else holding this off?

Dschee commented:
Just rebased to solve merge conflicts.

Dschee commented:
Just rebased this branch onto current master.

Dschee commented:
@jpsim I just rebased this, too. Thanks for merging the others!

For an example on how we configure these rules, see [here](https://github.com/JamitLabs/ProjLintTemplates/blob/master/iOS/SwiftLint.stencil#L95) and [there](https://github.com/JamitLabs/ProjLintTemplates/blob/master/iOS/SwiftLint.stencil#L107).

jpsim commented on Source/SwiftLintFramework/Rules/Style/TypeContentsOrderRule.swift:
```diff
@@ -0,0 +1,142 @@
+        var violations =  [StyleViolation]()
+
+        var lastMatchingIndex = -1
+        print(configuration.order)
```
rogue print call

Dschee commented:
I just rebased this to resolve merge conflicts. @jpsim Would you mind taking another look? I fixed your small comment a week ago, do you have any other reservations or can we merge now (once the CI has passed)?

Dschee commented:
Just rebased this branch to solve merge conflicts. Any idea when this might be merged @jpsim?

Dschee commented:
@marcelofabri Would you mind taking a look at this, too when you have time? I know this is a new rule and takes more time to understand. But:

* it's waiting for 6 months now and was rebased 5 times now to keep it up-to-date
* @jpsim already had a cursory look without telling me this is a bad idea
* we are using the rule via my fork in several projects at our company and it works well
* you can see it in action or play with it in the open source projects [BartyCrouch](https://github.com/Flinesoft/BartyCrouch), [HandySwift](https://github.com/Flinesoft/HandySwift) and [HandyUIKit](https://github.com/Flinesoft/HandyUIKit)
* it's opt-in and therefore shouldn't make any troubles with existing projects
* people seem to wait for it since it's the [most upvoted](https://github.com/realm/SwiftLint/pulls?q=is%3Apr+is%3Aopen+sort%3Areactions-%2B1-desc) PR right now

Would be glad to see this merged soon. It opens new ways to enforce Swift style & conventions, which is what this tool is all about.

Dschee commented:
@marcelofabri @jpsim I don't want to rush it and also don't have a good feeling in merging my own PR, but since you don't seem to have time for further reviewing this and it's opt-in and upvoted multiple times and is tested thouroughly, I'd go for it unless you express any complaints within the next say 10 days.

rumnat commented:
Thanks for this rule. How to setup? I can't open the configure examples

Dschee commented:
@rumnat Oh yes, the links changed, I just updated them, they should work again!
(The code is currently commented out since this PR is still not merged, will un-comment once this is merged and a new release was made.)

Dschee commented:
Okay, just as I announced a month ago (and I waited way longer than 10 days), since there was no feedback whatsoever since then and due to the reasons explained [here](https://github.com/realm/SwiftLint/pull/2296#issuecomment-463150590), I'm merging this without additional feedback from other contributors once CI is passed.

rumnat commented:
I'd like to have the following structure:
1) enums
2) protocols
3) Main Class
  - outlets
  - private field
  - public fields
  - overriding methods
  - public methods 
4) Private extension of main class
5) Rest of Extensions for each protocol implementation or separate logic

Can I reach it using your rule? Thanks in advance

Dschee commented:
You can't differentiate between the types as of now (e.g. enums vs. protocols, both would belong to `supporting_type`) also not between `private` and `public`, but try this with the `file_types_order` rule:

```yaml
file_types_order:
  order:
    - supporting_type
    - main_type
    - extension
```

Note that this is actually the default, so you could also just turn on the rule and it should be like this.

Regarding the type contents, what you want is currently not possible as you can't differentiate between `private` and `public` accessor, also you currently can't detect if a method is overriding or not. Those would all be even more detailed and could be added as an improvement to this rule. You can define the kinds of contents though, to me it looks like this might fit your preferences (assuming you first prefer type functions, then object functions):

```yaml
type_contents_order:
  order:
    - case
    - [type_alias, associated_type, subtype]
    - ib_outlet
    - [type_property, ib_inspectable, instance_property]
    - initializer
    - type_method
    - view_life_cycle_method
    - [ib_action, other_method]
    - subscript
```

Feel free to create a issues which request an additional rulse named something like `visibility_modifier_order` and `override_order` where we could enforce a specific modifier order given the type is the same. E.g. specifying something like this:

```
visibility_modifier_order:
  order:
    - [public, open]
    - [internal, implicit]
    - [private, fileprivate]

override_order:
  override_top: true
```

piobyz commented:
> For an example on how we configure these rules, see [here](https://github.com/JamitLabs/ProjLintTemplates/blob/master/JamitLabs/App/SwiftLint.stencil#L97)

Wow, what a great source of Custom Rules, thanks @Dschee

jpsim commented:
Thanks for all your work on this PR and really stepping up as a contributor and maintainer of this project overall @Dschee! I especially appreciate how thorough you were in adding tests and considering edges cases here.

> Okay, just as I announced a month ago (and I waited way longer than 10 days), since there was no feedback whatsoever since then and due to the reasons explained [here](https://github.com/realm/SwiftLint/pull/2296#issuecomment-463150590), I'm merging this without additional feedback from other contributors once CI passes.

I understand how frustrating it must be to not get feedback on your work for very long stretches of time. You did the right thing merging this. I'm sorry for not being very involved in SwiftLint on a regular basis any more 😕.

SwiftLint doesn't have a single gatekeeper (I'm sorry if I ever gave that impression). There are over a dozen people with write access to this git repository and **any one of those contributors should feel empowered to merge PRs, cut releases and help review other PRs/issues**.

I hope you'll keep contributing to this project as you've really made it so much better by your contributions so far.

Dschee commented:
Thank you for those kind words @jpsim and also for approving my self-merge. I still feel bad about it since usually I‘m the guy stressing how important a review is, but I felt different in this case. Thank you for your understanding!

I can understand that maintaining open source software - especially when there are so many users with different needs - can become hard at times. That‘s why I tried to help out when I had some time left over. Let‘s keep improving SwiftLint all together!

samrayner commented:
Thanks for this awesome addition @Dschee!

I'm getting one false positive. Files that only contain extensions are being flagged with:

> File Types Order Violation: An 'extension' should not be placed amongst the file type(s) 'main_type'. (file_types_order)

Dschee commented:
@samrayner Could you please open an issue with that false positive so we can track it and fix it?

Also, please provide a code example where this happens so we can properly reproduce it.

samrayner commented:
@Dschee https://github.com/realm/SwiftLint/issues/2749 Sorry for the delay!

Lutzifer commented:
great rule. How do we get this to autocorrect?

rogerluan commented:
Just dropping a line to show my appreciation over this rule! Absolutely great addition, can't wait to use it! 🚀
